src/main/java/net/fabricmc/loom/configuration/processors/JarProcessorManager.java

```diff
@@ -47,13 +65,56 @@ public boolean isInvalid(File file) {
 			return true;
 		}
 
+		String jarProcessorHash = getJarProcessorHash();
+
+		try (JarFile jar = new JarFile(file)) {
+			Attributes attributes = jar.getManifest().getMainAttributes();
+
+			if (!jarProcessorHash.equals(attributes.getValue(JAR_PROCESSOR_HASH_ATTRIBUTE))) {
+				return true;
+			}
+		} catch (IOException e) {
+			throw new UncheckedIOException("Could not check jar manifest of " + file, e);
+		}
+
 		return jarProcessors.stream().anyMatch(jarProcessor -> jarProcessor.isInvalid(file));
 	}
 
+	private String getJarProcessorHash() {
+		String jarProcessorIds = jarProcessors.stream()
+				.map(JarProcessor::getId)
+				.sorted()
+				.collect(Collectors.joining(";"));
+
+		try {
+			return CharSource.wrap(jarProcessorIds)
+					.asByteSource(StandardCharsets.UTF_8)
+					.hash(Hashing.sha256())
+					.toString().substring(0, 12);
```

Comment:
Hmm, might as well use the full 64-character hash here as a few extra bytes don't matter in the manifest (and there's no need to substring then)

Reply:
```suggestion
					.toString();
```